assets/blocks/course-outline/store.js

```diff
@@ -118,23 +131,53 @@ export const COURSE_STORE = 'sensei/course-structure';
  * Register course structure store and subscribe to block editor save.
  */
 const registerCourseStructureStore = () => {
-	subscribe( function saveStructureOnPostSave() {
-		const editor = select( 'core/editor' );
+	/**
+	 * Save course structure.
+	 *
+	 * @param {boolean} isSavingPost     Whether the post is saving.
+	 * @param {boolean} shouldResavePost Whether the post should resave.
+	 */
+	const saveCourseStructure = ( isSavingPost, shouldResavePost ) => {
+		const isSaveCalled = select( COURSE_STORE ).isSaveCalled();
 
-		if ( ! editor ) return;
+		// Make sure to run the save once after every post saving.
+		if ( isSavingPost ) {
+			if ( isSaveCalled ) {
+				return;
```

Comment:
These subscribe functions are pretty crucial, so we should clean it up as much as possible. 
I think we can keep the `saveCalled` tracking out of the structure store, as it's really just relevant for this subscribe handling. 
What do you think of separating the part that figures out the post saving events completely, providing a `startSave` and `finishSave` that are called once when the post starts/finishes saving:

```js
// Set to true when savings starts, and false when it ends
let postSaving = false;

subscribe( function handlePostSaving() {
	const editor = select( 'core/editor' );

	if ( ! editor ) {
		return;
	}

	const isSavingPost =
		editor.isSavingPost() && ! editor.isAutosavingPost();
	
	// First update where post is saving => start hook
	if ( ! postSaving && isSavingPost ) {
		postSaving = true;
		startSave();
		return;
	}
	
	// First update where post is no longer saving => finish hook
	if ( postSaving && ! isSavingPost ) {
		postSaving = false;
		finishSave();
		return;
	}
} );
``` 

Then the `startSave` and `finishSave` functions just handle anything that needs to happen at those points:
```js
	const startSave = () => {
		const shouldSave = select( COURSE_STORE ).shouldSave();

		// Clear error notices.
		dispatch( 'core/notices' ).removeNotice( 'course-outline-save-error' );

		if ( shouldSave ) {
			dispatch( COURSE_STORE ).save();
		}
	};

	const finishSave = () => {
		// Save the post again if the blocks were updated.
		const shouldResavePost = select( COURSE_STORE ).shouldResavePost();
		if ( shouldResavePost ) {
			dispatch( 'core/editor' ).savePost();
		}
	};
```

Reply:
Good idea about handle in a local variable and the `postSaving` naming! Refactored here: https://github.com/Automattic/sensei/pull/3707/commits/fb6e467ba4aa5581e5144dad8cd9279d9e1455b4